### PR TITLE
Bug 1800425: More appropriate annotation name

### PR DESCRIPTION
**Pull request by beekhof**
Since MHC has no control over how the machine is remediated, it would be
better not to imply that it will (only) be via reboot.

### CODE REVIEW

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ import (
 
 const (
 	machineAnnotationKey          = "machine.openshift.io/machine"
-	machineRebootAnnotationKey    = "healthchecking.openshift.io/machine-remediation-reboot"
```
Comment (enxebre) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: can you also change `healthchecking.openshift.io` -> `machine.openshift.io`

Comment (enxebre) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: wouldn't you want to chance the variable name as well? and the func `remediationStrategyReboot` name and its events and logging?

Comment (beekhof) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: good point, I'll do that too

Comment (beekhof) [reply pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: also `healthchecking.openshift.io/strategy` -> `machine.openshift.io/remediation-strategy` ?

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 2]
```diff
@@ -471,24 +471,24 @@ func (t *target) remediate(r *ReconcileMachineHealthCheck) error {
...
+	glog.Infof("Machine %s has been unhealthy for too long, adding external annotation", t.Machine.Name)
+	t.Node.Annotations[machineExternalAnnotationKey] = ""
 	if err := r.client.Update(context.TODO(), t.Node); err != nil {
 		r.recorder.Eventf(
 			&t.Machine,
 			corev1.EventTypeWarning,
-			EventRebootAnnotationFailed,
-			"Requesting reboot of node associated with machine %v failed: %v",
+			EventExternalAnnotationFailed,
+			"Requesting external of node associated with machine %v failed: %v",
```
Comment (markmc) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: s/external of node/external remediation of node/

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 3]
```diff
@@ -497,8 +497,8 @@ func (t *target) remediationStrategyReboot(r *ReconcileMachineHealthCheck) error
 	r.recorder.Eventf(
 		&t.Machine,
 		corev1.EventTypeNormal,
-		EventRebootAnnotationAdded,
-		"Requesting reboot of node associated with machine %v",
+		EventExternalAnnotationAdded,
+		"Requesting external of node associated with machine %v",
```
Comment (markmc) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]: Same here

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 4]
```diff
@@ -28,14 +28,14 @@ import (
 
 const (
 	machineAnnotationKey          = "machine.openshift.io/machine"
-	machineRebootAnnotationKey    = "healthchecking.openshift.io/machine-remediation-reboot"
+	machineExternalAnnotationKey  = "machine.openshift.io/external-remediation"
```
Comment (beekhof) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]:
```suggestion
	machineExternalAnnotationKey  = "host.metal3.io/external-remediation"
```

**Comment by beekhof** [comment]
/bugzilla refresh

@enxebre I was going to wait until the other PR merged and then rebase this one

**Comment by bison** [comment]
I agree with the intent here, but now the strategy name remains generic (reboot -> external) while the annotation is provider-specific: `host.metal3.io/external-remediation`

Could we make the strategy name match more closely?

Also, I forget, is this still feature gated in a way that blocks upgrades? If not, even if it's still tech-preview, can we update it like this? If I update an existing cluster that uses the `reboot` strategy, suddenly that's not recognized and I get the "delete" strategy with no warning.

**Comment by beekhof** [comment]
Rebased now that #475 has been merged.

> I agree with the intent here, but now the strategy name remains generic (reboot -> external) while the annotation is provider-specific: `host.metal3.io/external-remediation`
> 
> Could we make the strategy name match more closely?

Are we just talking about changing
``remediationStrategyExternal   = mapiv1.RemediationStrategyType("external”)``
to something like
``remediationStrategyExternal   = mapiv1.RemediationStrategyType(“baremetal-external”)``
to indicate that this only works for baremetal?

Otherwise, the challenge of using `machine.openshift.io/external-remediation` is that metal3 see's `openshift.io` in the annotation name and thinks "that's openshift specific, we shouldn't be implementing that".  

I don't think we can both have `machine.openshift.io/external-remediation` as a platform agnostic annotation name, _and_ push the implementation onto providers that need to run elsewhere.

> Also, I forget, is this still feature gated in a way that blocks upgrades? If not, even if it's still tech-preview, can we update it like this? If I update an existing cluster that uses the `reboot` strategy, suddenly that's not recognized and I get the "delete" strategy with no warning.

I don't believe it is feature gated in 4.3, but we've also never shipped anything that depended these annotations.  That does not exclude the possibility that customers have dug into the MAO internals and implemented something - but are annotations held to the same versioning standards as CRDs? I thought that was part of the attraction of using annotations when the "final" interface is still an open question.

**Comment by beekhof** [comment]
Resolved goof when merging with #475
Remaining test failures seem to be from etcd-operator

File: pkg/controller/machinehealthcheck/machinehealthcheck_controller.go [hunk 5]
```diff
@@ -28,14 +28,14 @@ import (
...
+	machineExternalAnnotationKey  = "host.metal3.io/external-remediation"
 	ownerControllerKind           = "MachineSet"
 	nodeMasterLabel               = "node-role.kubernetes.io/master"
 	machineRoleLabel              = "machine.openshift.io/cluster-api-machine-role"
 	machineMasterRole             = "master"
 	machinePhaseFailed            = "Failed"
-	remediationStrategyAnnotation = "healthchecking.openshift.io/strategy"
-	remediationStrategyReboot     = mapiv1.RemediationStrategyType("reboot")
+	remediationStrategyAnnotation = "machine.openshift.io/remediation-strategy"
+	remediationStrategyExternal   = mapiv1.RemediationStrategyType("external")
```
Comment (beekhof) [inline pkg/controller/machinehealthcheck/machinehealthcheck_controller.go]:
```suggestion
	remediationStrategyExternal   = mapiv1.RemediationStrategyType("external-baremetal")
```

**Comment by bison** [comment]
> Are we just talking about changing
> remediationStrategyExternal = mapiv1.RemediationStrategyType("external”)
> to something like
> remediationStrategyExternal = mapiv1.RemediationStrategyType(“baremetal-external”)
> to indicate that this only works for baremetal?

Yeah, sorry, that wasn't very clear. More or less what you described, yeah. If different strategies result in provider specific annotations going onto Machine objects, I just think the `RemediationStrategyType` string should reflect the provider in some way. So `baremetal-external` or `metal3` or whatever would be fine.

> I don't believe it is feature gated in 4.3, but we've also never shipped anything that depended these annotations. That does not exclude the possibility that customers have dug into the MAO internals and implemented something - but are annotations held to the same versioning standards as CRDs? I thought that was part of the attraction of using annotations when the "final" interface is still an open question.

Fair enough. I think you're probably right. I mostly wanted to bring up that the current behavior for an unknown strategy is to just silently fallback to "delete the Machine", which I think is a separate issue. I'll file a bug on that.

**Comment by beekhof** [comment]
@bison happy enough with the change to give lgtm?

**Comment by beekhof** [comment]
@enxebre what do you think, can we target this for 4.4?

**Pull request merged**